subprojects/plugins/src/main/java/org/gradle/api/tasks/Upload.java

```diff
@@ -16,143 +16,33 @@
 
 package org.gradle.api.tasks;
 
-import groovy.lang.Closure;
-import groovy.lang.DelegatesTo;
-import org.gradle.api.Action;
+import org.gradle.api.InvalidUserCodeException;
 import org.gradle.api.artifacts.Configuration;
-import org.gradle.api.artifacts.PublishException;
-import org.gradle.api.artifacts.dsl.RepositoryHandler;
-import org.gradle.api.file.FileCollection;
 import org.gradle.api.internal.ConventionTask;
-import org.gradle.api.internal.artifacts.ArtifactPublicationServices;
-import org.gradle.api.internal.artifacts.ArtifactPublisher;
-import org.gradle.api.internal.artifacts.Module;
-import org.gradle.api.internal.artifacts.configurations.ConfigurationInternal;
-import org.gradle.api.internal.artifacts.repositories.PublicationAwareRepository;
-import org.gradle.internal.Transformers;
-import org.gradle.internal.deprecation.DeprecationLogger;
-import org.gradle.util.internal.ConfigureUtil;
+import org.gradle.api.internal.DocumentationRegistry;
 import org.gradle.work.DisableCachingByDefault;
 
-import javax.annotation.Nullable;
 import javax.inject.Inject;
-import java.io.File;
-import java.util.List;
-
-import static org.gradle.util.internal.CollectionUtils.collect;
 
 /**
  * Uploads the artifacts of a {@link Configuration} to a set of repositories.
  *
  * @deprecated This class is scheduled for removal in Gradle 8.0. To upload artifacts, use the maven-publish plugin instead.
  */
-@Deprecated
+@Deprecated // TODO:Finalize Upload Removal - Issue #21439
 @DisableCachingByDefault(because = "Produces no cacheable output")
 public class Upload extends ConventionTask {
-
-    private Configuration configuration;
-    private boolean uploadDescriptor;
-    private File descriptorDestination;
-    private RepositoryHandler repositories;
+    private final DocumentationRegistry documentationRegistry;
```

Comment:
You can do this as an `@Inject abstract DocumentationRegistry getDocumentationRegistry();` and then you don't need the new constructor.

You'll have to make the class abstract too, but I think that's OK.

Reply:
I though we might want to make it `abstract` to avoid breaking other plugins that might be creating instances, and figured it was going to be removed anyways.  I'll change it.